File: 2021/07/28/PathSum.java
/**
 * https://leetcode.com/problems/path-sum/
 * @author Andrew Jarombek
 * @since 7/28/2021
 */

import java.util.HashSet;

class PathSum {

    public static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {}

        TreeNode(int val) { this.val = val; }

        TreeNode(int val, TreeNode left, TreeNode right) {
           this.val = val;
           this.left = left;
           this.right = right;
        }
    }
    
    public static boolean hasPathSum(TreeNode root, int targetSum) {
        if (root == null) {
            return false;
        }

        HashSet<TreeNode> treeNodes = new HashSet<>();
        treeNodes.add(root);

        while (!treeNodes.isEmpty()) {
            HashSet<TreeNode> newTreeNodes = new HashSet<>();

            for (TreeNode node: treeNodes) {
                if (node.val == targetSum && node.left == null && node.right == null) {
                    return true;
                }

                if (node.left != null) {
                    node.left.val = node.left.val + node.val;
                    newTreeNodes.add(node.left);
                }

                if (node.right != null) {
                    node.right.val = node.right.val + node.val;
                    newTreeNodes.add(node.right);
                }
            }

            treeNodes = newTreeNodes;
        }

        return false;
    }

    public static void main(String[] args) {
        TreeNode node = new TreeNode(2, new TreeNode(1), new TreeNode(3));
        assert !PathSum.hasPathSum(node, 5);
        assert PathSum.hasPathSum(node, 3);
    }
}

File: 2022/11/14/cpp/pathSum.cpp
// https://leetcode.com/problems/path-sum/
// Author: Andrew Jarombek
// Date: 11/14/2022

#include <cassert>
#include <algorithm>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

bool hasPathSum(TreeNode* root, int targetSum);
bool dfs(TreeNode* root, int currentSum, int targetSum);

bool hasPathSum(TreeNode* root, int targetSum) {
    if (root == nullptr)
        return false;

    return dfs(root, 0, targetSum);
}

bool dfs(TreeNode* root, int currentSum, int targetSum) {
    if (root->left == nullptr && root->right == nullptr)
        return currentSum + root->val == targetSum;

    bool left = false;
    bool right = false;

    if (root->left != nullptr) {
        left = dfs(root->left, currentSum + root->val, targetSum);
    }

    if (root->right != nullptr) {
        right = dfs(root->right, currentSum + root->val, targetSum);
    }

    return left || right;
}

int main() {
    auto* tree = new TreeNode(1, new TreeNode(2), new TreeNode(3));
    assert(!hasPathSum(tree, 1));
    assert(!hasPathSum(tree, 2));
    assert(hasPathSum(tree, 3));
    assert(hasPathSum(tree, 4));
}